### DISCUSSION TITLE
Include errors

### DISCUSSION THREAD

**Discussion by andrey1887** [idaholab/moose#25775] [Q&A General] [answered]
Hello. I try to create a custom aux kernel and currently i start with two files : 
/home/and/projects/thin_walled/include/auxkernels/LocalAxisVectorAux.h
``` 
#pragma once
#include "Kernel.h" 
```

and
/home/and/projects/thin_walled/src/auxkernels/LocalAxisVectorAux.C
```
#include "LocalAxisVectorAux.h"
registerMooseObject("LocalAxisVector", LocalAxisVectorAux);
```
But i face such problem with includes in these both files:
```
#include errors detected. Consider updating your compile_commands.json or includePath. 
Squiggles are disabled for this translation unit 
(/home/and/projects/thin_walled/src/auxkernels/LocalAxisVectorAux.C).C/C++(1696)
```
My c_cpp_properties.json:
```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}",
                "${workspaceFolder}/build/header_symlinks",
                "${workspaceFolder}/../moose/framework/build/header_symlinks",
                "${workspaceFolder}/../moose/modules/module_loader/build/header_symlinks",
                "${workspaceFolder}/../moose/framework/contrib/boost/include",
                "${workspaceFolder}/../moose/framework/contrib/hit",
                "${HOME}/mambaforge3/envs/moose/include",
                "${HOME}/mambaforge3/envs/moose/libmesh/include",
                "${workspaceFolder}/../moose/framework/contrib/json/include"
            ],
            "defines": [],
            "compilerPath": "${HOME}/mambaforge3/envs/moose/bin/mpicxx",
            "cStandard": "c11",
            "cppStandard": "c++17",
            "intelliSenseMode": "clang-x64",
            "compileCommands": "${workspaceFolder}/compile_commands.json",
            "browse": {
                "path": [
                    "${workspaceFolder}",
                    "${workspaceFolder}/build/header_symlinks",
                    "${workspaceFolder}/../moose/framework/build/header_symlinks",
                    "${workspaceFolder}/../moose/modules/module_loader/build/header_symlinks",
                    "${workspaceFolder}/../moose/framework/contrib/boost/include",
                    "${workspaceFolder}/../moose/framework/contrib/hit",
                    "${HOME}/mambaforge3/envs/moose/include",
                    "${HOME}/mambaforge3/envs/moose/libmesh/include"
                ]
            }
        }
    ],
    "version": 4
}

```
and settings.json
```
{
    "workbench.colorTheme": "Atom One Dark",
    "gitlens.hovers.currentLine.over": "line",
    "[cpp]": {
        "editor.defaultFormatter": "ms-vscode.cpptools"
    },
    "telemetry.telemetryLevel": "off",
    "C_Cpp.default.cppStandard": "c++17",
    "C_Cpp.default.includePath": [
        "${workspaceFolder}/**",
        "${env.CONDA_PREFIX}/**",
        "/usr/include/**"
    ],
    "files.associations": {
        "*.C": "cpp",
        "*.h": "cpp"
    },
    "files.trimTrailingWhitespace": true,
    "editor.tabSize": 2,
    "editor.formatOnSave": true,
    "files.insertFinalNewline": true
}
```
Help me please with this issue. What i doing wrong?

**Comment by GiudGiud** [comment]
If it s all there is in those files, that won't compile

You need to define a full blown class before being able to register it

**Reply by GiudGiud** [reply, marked as answer]
I see `workspaceFolder` in the list of folders for the includes. Did you open vscode in the app's folder? 
```
cd projects/<app_folder>
code ./
```

**Reply by andrey1887** [reply]
VSCode is opened in folder thin_walled:
![изображение](https://github.com/idaholab/moose/assets/114737834/3ef8f0f4-4fd6-4218-bb9a-1fe8a6f93d8e)

**Reply by andrey1887** [reply]
What is a right order of app development which allow to avoid such errors?

**Reply by andrey1887** [reply]
@GiudGiud explain me please, what additional directories and  files i need for custom kernel development?

**Reply by GiudGiud** [reply]
these folders are fine. can you compile on the command line following our getting started instructions?

**Reply by andrey1887** [reply]
I try to compile and got following message:
![изображение](https://github.com/idaholab/moose/assets/114737834/3b169928-44d3-435b-94e6-a03e2c115d8a)

**Reply by GiudGiud** [reply]
does that file exist in MOOSE?

**Reply by andrey1887** [reply]
No. I thought, that it must included from moose subfolders, isn't it?
![изображение](https://github.com/idaholab/moose/assets/114737834/cd994061-427e-40c5-9d4f-1707ab1dc288)

**Reply by GiudGiud** [reply]
where is it? VectorAuxKernel.h ? is it in moose/framework/include?

I dont see it

**Comment by andrey1887** [comment]
Yes you are right this file is absent. If i change include to this #include "AuxKernel.h" compilation run fine, but there are still errors in the includes:
 
![изображение](https://github.com/idaholab/moose/assets/114737834/14f98bed-9e22-4fce-b634-a7d3d8587c51)

Which file do I need to inherit from to create custom VectorAuxKernel, that you mensioned here https://github.com/idaholab/moose/discussions/25653#discussioncomment-7179458.

**Reply by GiudGiud** [reply]
add                 "${workspaceFolder}/include/*",
or                 "${workspaceFolder}/include",
to the browser_path in the json

**Reply by GiudGiud** [reply]
just include auxkernel.h
see in that header, the definition of a VectorAuxkernel
```
typedef AuxKernelTempl<RealVectorValue> VectorAuxKernel;
```

**Reply by andrey1887** [reply]
After adding:
![изображение](https://github.com/idaholab/moose/assets/114737834/5725cbb8-a929-43ea-8227-dc8080e4254d)
![изображение](https://github.com/idaholab/moose/assets/114737834/6c78d39a-363b-4917-be71-e730981de352)

**Reply by GiudGiud** [reply]
does it compile though

**Reply by andrey1887** [reply]
There were two wrong things supposedly:

1. redundant folder ![Безымянный](https://github.com/idaholab/moose/assets/114737834/7935ca22-ca6c-4f18-b643-0ddaa2949cd5)

2.And VSCode must be started this way: https://github.com/idaholab/moose/discussions/25775#discussioncomment-7327193


Thanks a lot for help.

**Comment by andrey1887** [comment]
Hello, @GiudGiud. After reinstallation of moose something went wrong. 

1. `mambaforge3` was changed to `miniforge`:
```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}",
                "${workspaceFolder}/build/header_symlinks",
                "${workspaceFolder}/../moose/framework/build/header_symlinks",
                "${workspaceFolder}/../moose/modules/module_loader/build/header_symlinks",
                "${workspaceFolder}/../moose/framework/contrib/boost/include",
                "${workspaceFolder}/../moose/framework/contrib/hit",
                "${HOME}/miniforge/envs/moose/include",
                "${HOME}/miniforge/envs/moose/libmesh/include",
                "${workspaceFolder}/../moose/framework/contrib/json/include",
                "${workspaceFolder}/include/**"
            ],
            "defines": [],
            "compilerPath": "${HOME}/miniforge/envs/moose/bin/mpicxx",
            "cStandard": "c11",
            "cppStandard": "c++17",
            "intelliSenseMode": "clang-x64",
            "compileCommands": "${workspaceFolder}/compile_commands.json",
            "browse": {
                "path": [
                    "${workspaceFolder}",
                    "${workspaceFolder}/build/header_symlinks",
                    "${workspaceFolder}/../moose/framework/build/header_symlinks",
                    "${workspaceFolder}/../moose/modules/module_loader/build/header_symlinks",
                    "${workspaceFolder}/../moose/framework/contrib/boost/include",
                    "${workspaceFolder}/../moose/framework/contrib/hit",
                    "${HOME}/miniforge/envs/moose/include",
                    "${HOME}/miniforge/envs/moose/libmesh/include",
                    "${workspaceFolder}/include/**"
                ]
            }
        }
    ],
    "version": 4
}
```

2. LocalAxisAux.h:
![изображение](https://github.com/idaholab/moose/assets/114737834/70dbab67-0925-46d0-813a-57f916c15adc)

3. LocalAxisAux.C
![изображение](https://github.com/idaholab/moose/assets/114737834/6e20d253-d4b3-40ef-9e7b-3c058f8ddb58)

4.After attemp to `make`:
```
(moose) and@pc:~/projects/thin_walled$ make
Checking if header needs updating: /home/and/projects/moose/framework/include/base/MooseRevision.h...
Compiling C++ (in opt mode) /home/and/projects/thin_walled/build/unity_src/auxkernels_Unity.C...
In file included from /home/and/projects/thin_walled/build/unity_src/auxkernels_Unity.C:2:
/home/and/projects/thin_walled/src/auxkernels/LocalAxisAux.C: In member function 'virtual libMesh::Real LocalAxisAux::computeValue()':
/home/and/projects/thin_walled/src/auxkernels/LocalAxisAux.C:37:33: error: no matching function for call to 'SubProblem::assembly(THREAD_ID&)'
   37 |       &_subproblem.assembly(_tid).getFE(FEType(), 1)->get_dxyzdxi();
      |                                 ^
In file included from /home/and/projects/moose/framework/build/header_symlinks/MooseVariableFE.h:14,
                 from /home/and/projects/moose/framework/build/header_symlinks/Assembly.h:15,
                 from /home/and/projects/thin_walled/src/auxkernels/LocalAxisAux.C:1,
                 from /home/and/projects/thin_walled/build/unity_src/auxkernels_Unity.C:2:
/home/and/projects/moose/framework/build/header_symlinks/SubProblem.h:310:22: note: candidate: 'virtual Assembly& SubProblem::assembly(THREAD_ID, unsigned int)'
  310 |   virtual Assembly & assembly(const THREAD_ID tid, const unsigned int nl_sys_num) = 0;
      |                      ^~~~~~~~
/home/and/projects/moose/framework/build/header_symlinks/SubProblem.h:310:22: note:   candidate expects 2 arguments, 1 provided
/home/and/projects/moose/framework/build/header_symlinks/SubProblem.h:311:28: note: candidate: 'virtual const Assembly& SubProblem::assembly(THREAD_ID, unsigned int) const'
  311 |   virtual const Assembly & assembly(const THREAD_ID tid, const unsigned int nl_sys_num) const = 0;
      |                            ^~~~~~~~
/home/and/projects/moose/framework/build/header_symlinks/SubProblem.h:311:28: note:   candidate expects 2 arguments, 1 provided
make: *** [/home/and/projects/moose/framework/build.mk:151: /home/and/projects/thin_walled/build/unity_src/auxkernels_Unity.x86_64-conda-linux-gnu.opt.lo] Ошибка 1

```
Help me please to resolve this issue.

**Comment by GiudGiud** [comment]
Hello

1 is normal. We switched from mambaforge to miniforge

2 & 3 are vscode 'warnings'. Essentially you ll need to add the path to wasp (moose-wasp is from a conda package right?) to the search page for vscode

4 is a change in APIs.
We no longer let you access the Assembly without specifying the system number.
So please specify the nonlinear system number in those calls:
```
&_subproblem.assembly(_tid, 0)....
```
most likely

Guillaume

**Reply by andrey1887** [reply]
Thank you, @GiudGiud . After adding `"${HOME}/miniforge/envs/moose/wasp/include"` in include path of `c_cpp_properties.json` a squiggles are gone. 

Related to `system number`:
In the `ComputreIncrementalBeamStrain.C` from which this line came in my `LocalAxisAux.C` it was changed to:
`&_subproblem.assembly(_tid, _nonlinear_sys.number()).getFE(FEType(), 1)->get_dxyzdxi();`

Should i use a new above variant with:
`_nonlinear_sys(_fe_problem.getNonlinearSystemBase(/*nl_sys_num=*/0))` in constructor

 or specify `0` as second argument as you show here: 
`&_subproblem.assembly(_tid, 0)....` is enough ?

**Reply by GiudGiud** [reply]
every single function call where it now requires a system number you have to input it.

I assume you are only using a single non linear system so 0 is a good value for now.
When there is a `_nonlinear_system` member attribute, `_nonlinear_system.number()` is a better choice for sure

**Reply by andrey1887** [reply]
Thank you, @GiudGiud . When i try to pass `_nonlinear_sys(_fe_problem.getNonlinearSystemBase(/*nl_sys_num=*/0))` as parameter to constructor i get the next:
![изображение](https://github.com/idaholab/moose/assets/114737834/7753d659-ef9c-42fd-81c5-65aa533a825a)
 How it can be resolved?

**Reply by GiudGiud** [reply]
your class does not have a _fe_problem attribute. 

but every auxkernel has these
```
  /// Subproblem this kernel is part of
  SubProblem & _subproblem;
  /// System this kernel is part of
  SystemBase & _sys;
  SystemBase & _nl_sys;
  AuxiliarySystem & _aux_sys;
```

so maybe just use _nl_sys ?

**Reply by andrey1887** [reply]
@GiudGiud, did you mean this:

```&_subproblem.assembly(_tid, _nl_sys.number()).getFE(FEType(), 1)->get_dxyzdxi();``` ?

Before this "SystemBase.h" was included , line ```_nonlinear_sys(_fe_problem.getNonlinearSystemBase(/*nl_sys_num=*/0))``` was deleted. No other changes. Is it right ?
![изображение](https://github.com/idaholab/moose/assets/114737834/b733389c-6572-4fa5-be8f-3105f2537c51)

**Reply by GiudGiud** [reply]
This looks fine to me but I have not compiled it myself

**Reply by andrey1887** [reply]
Thank, you, @GiudGiud . Compilation completed successfully.
